src/ray/gcs/job_state_accessor.h

```diff
@@ -0,0 +1,71 @@
+#ifndef RAY_GCS_JOB_STATE_ACCESSOR_H
+#define RAY_GCS_JOB_STATE_ACCESSOR_H
+
+#include "ray/common/id.h"
+#include "ray/gcs/callback.h"
+#include "ray/gcs/subscription_executor.h"
+#include "ray/gcs/tables.h"
+
+namespace ray {
+
+namespace gcs {
+
+class RedisGcsClient;
+
+/// \class JobStateAccessor
+/// JobStateAccessor class encapsulates the implementation details of
+/// reading or writing or subscribing of job's information (immutable fields which
+/// determined at submission time, and mutable fields which determined at runtime).
+class JobStateAccessor {
+ public:
+  explicit JobStateAccessor(RedisGcsClient &client_impl);
+
+  ~JobStateAccessor() {}
+
+  /// Register a job to GCS asynchronously.
+  ///
+  /// \param data_ptr The job that will be registered to GCS.
+  /// \param callback Callback that will be called after job has been registered
+  /// to GCS.
+  /// \return Status
+  Status AsyncRegister(const std::shared_ptr<JobTableData> &data_ptr,
+                       const StatusCallback &callback);
+
+  /// Update dynamic states of job in GCS asynchronously.
+  ///
+  /// \param data_ptr The job that will be updated to GCS.
+  /// \param callback Callback that will be called after update finishes.
+  /// \return Status
+  Status AsyncUpdate(const std::shared_ptr<JobTableData> &data_ptr,
+                     const StatusCallback &callback);
+
+  /// Subscribe to any update operations of jobs.
+  ///
+  /// \param subscribe Callback that will be called each time when an job is registered
+  /// or updated.
+  /// \param done Callback that will be called when subscription is complete.
+  /// \return Status
+  Status AsyncSubscribeAll(const SubscribeCallback<JobID, JobTableData> &subscribe,
```

Comment:
Same reason as above, this can be `AsyncSubscribeToDeadJobs`.

Reply:
done